lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -893,25 +896,133 @@ def default_ami_for_region(region, criteria = {})
 
       winrm_options = {
-        :user => machine_spec.reference['winrm_username'] || 'Administrator',
-        :pass => decrypted_password,
-        :disable_sspi => true,
-        :basic_auth_only => true
+        user: username,
+        pass: password,
+        disable_sspi: disable_sspi,
+        basic_auth_only: basic_auth_only,
+        no_ssl_peer_verification: no_ssl_peer_verification,
       }
 
+      if no_ssl_peer_verification or type != :ssl
+        # we won't verify certs
+      elsif machine_spec.reference[:winrm_ssl_cert]
+        # we have stored the cert
+      else
+        # we need to retrieve the cert and verify it by connecting just to
+        # retrieve the ssl certificate and compare it to what we see in the
+        # console logs
+        console_lines = Base64.decode64(instance.console_output.data.output).lines
+        noverify_peer_context = OpenSSL::SSL::SSLContext.new
+        noverify_peer_context.verify_mode = OpenSSL::SSL::VERIFY_NONE
+        tcp_connection = TCPSocket.new(instance.private_ip_address, '5986')
+        shady_ssl_connection = OpenSSL::SSL::SSLSocket.new(tcp_connection, noverify_peer_context)
+        shady_ssl_connection.connect
+        shady_ssl_connection.peer_cert_chain
+        winrm_cert = shady_ssl_connection.peer_cert_chain.first
+
+        rdp_thumbprint = console_lines.grep(
+          /RDPCERTIFICATE-THUMBPRINT/)[-1].split(': ').last.chomp
+        rdp_subject = console_lines.grep(
+          /RDPCERTIFICATE-SUBJECTNAME/)[-1].split(': ').last.chomp
+        winrm_subject = winrm_cert.subject.to_s.split('=').last.upcase
+        winrm_thumbprint=OpenSSL::Digest::SHA1.new(winrm_cert.to_der).to_s.upcase
+
+        if rdp_subject != winrm_subject or rdp_thumbprint != winrm_thumbprint
+          Chef::Log.fatal "Winrm ssl port certificate differs from rdp console logs"
+        end
+        machine_spec.reference[:winrm_ssl_subject]=winrm_subject
+        machine_spec.reference[:winrm_ssl_thumbprint]=winrm_thumbprint
+        machine_spec.reference[:winrm_ssl_cert]=winrm_cert.to_pem
+      end
+
+      # If we have a cert, use it for verification
+      if machine_spec.reference[:winrm_ssl_cert]
+        FileUtils.mkdir_p(Chef::Config.trusted_certs_dir)
+        filename = File.join(Chef::Config.trusted_certs_dir, "#{machine_spec.name}.crt")
+        if File.exists?(filename)
+          Chef::Log.warn("Existing cert for #{winrm_subject} in #{filename}")
+        else
+          Chef::Log.warn("Adding certificate for #{winrm_subject} in #{filename}")
+          File.open(filename, File::CREAT|File::TRUNC|File::RDWR, 0644) do |f|
+            f.print(machine_spec.reference[:winrm_ssl_cert])
+          end
+        end
+        winrm_options[:ca_trust_path] = filename
+      end
       Chef::Provisioning::Transport::WinRM.new("#{endpoint}", type, winrm_options, {})
+      binding.pry
+      # in order for ssl to work, we need to tell ssl that it's ok that ssl_subject doesn't
```

Comment:
Never worked with certs in the WinRM gem, just with powershell from windows hosts. I'll have a look though today.